timemachine/cpp/src/kernels/k_nonbonded_common.cuh

```diff
@@ -12,23 +12,155 @@ static const int NONBONDED_KERNEL_THREADS_PER_BLOCK = 256;
+    double dij4 = dij2 * dij2;
+    double dij7 = dij4 * dij2 * dij;
+    double dij8 = dij4 * dij4;
+
+    // arg = 0.5 * pi * (dij / cutoff)^8
+    double dij_k8 = dij8 * k8;
+    double arg = 0.5 * pi * dij_k8;
+
+    double sin_arg;
+    double cos_arg;
+    sincos(arg, &sin_arg, &cos_arg);
+
+    double cos_arg2 = cos_arg * cos_arg;
+
+    return -12 * pi * dij7 * sin_arg * cos_arg2 * k8;
+}
+
+double __device__ __forceinline__ d_erfc_beta_r_dr(double beta, double dij) {
+    // -2 beta exp(-(beta dij)^2) / sqrt(pi)
+    double beta_dij = beta * dij;
     double exp_beta_dij_2 = exp(-beta_dij * beta_dij);
-    erfc_beta_dij = erfc(beta_dij);
-    return -inv_d2ij * (static_cast<double>(TWO_OVER_SQRT_PI) * beta_dij * exp_beta_dij_2 + erfc_beta_dij);
+    return -static_cast<double>(TWO_OVER_SQRT_PI) * beta * exp_beta_dij_2;
 }
 
-float __device__ __forceinline__ real_es_factor(float real_beta, float dij, float inv_d2ij, float &erfc_beta_dij) {
-    float beta_dij = real_beta * dij;
-    // max ulp error is: 2 + floor(abs(1.16 * x))
-    float exp_beta_dij_2 = __expf(-beta_dij * beta_dij);
-    // 5th order gaussian polynomial approximation, we need the exp(-x^2) anyways for the chain rule
+double __device__ __forceinline__
+real_es_factor(double real_beta, double dij, double inv_dij, double inv_d2ij, double &damping_factor) {
+    double beta_dij = real_beta * dij;
+    double erfc_beta_dij = erfc(beta_dij);
+
+    // write erfc(beta * dij) * switch_fn(dij) into damping_factor
+    damping_factor = erfc_beta_dij * switch_fn(dij);
+
+    double dsdr = d_switch_fn_dr(dij);
+    double debd = d_erfc_beta_r_dr(real_beta, dij);
+    double sr = switch_fn(dij);
+
+    // chain rule
+    double damping_factor_prime = (erfc_beta_dij * dsdr) + (debd * sr);
+    double d_es_dr = damping_factor_prime * inv_dij - damping_factor * inv_d2ij;
+    return d_es_dr;
+}
+
+// f32 code path merges (1) switch_fn and its deriv into switch_fn_and_deriv
+// and (2) a fast erfc approximation and its deriv into fast_erfc_and_deriv
+float __device__ __forceinline__ switch_fn_and_deriv(float dij, float *dsdr) {
+
+    // constants
+    constexpr float cutoff = 1.2;
+    constexpr float pi = static_cast<float>(PI);
+    constexpr float inv_cutoff = 1 / cutoff;
+
+    // cutoff^-1
+    constexpr float k2 = inv_cutoff * inv_cutoff;
+    constexpr float k4 = k2 * k2;
+    constexpr float k8 = k4 * k4;
+
+    // dij^7, dij^8
+    float dij2 = dij * dij;
+    float dij4 = dij2 * dij2;
+    float dij7 = dij4 * dij2 * dij;
+    float dij8 = dij4 * dij4;
+
+    // arg = (dij/cutoff)^8 * pi/2
+    float dij_k8 = dij8 * k8;
+    float arg = 0.5 * pi * dij_k8;
+
+    // sin(arg), cos(arg)
+    float sin_arg;
+    float cos_arg;
+    __sincosf(arg, &sin_arg, &cos_arg);
+
+    // exponentiation
+    float cos_arg2 = cos_arg * cos_arg;
+    float cos_arg3 = cos_arg2 * cos_arg;
+
+    // write d switch_fn d r
+    dsdr[0] = -12 * pi * dij7 * sin_arg * cos_arg2 * k8;
```

Comment:
constexpr for -12 * pi?

Reply:
Good call -- added in https://github.com/proteneer/timemachine/pull/1290/commits/92c16263e3706cde4244e136370fb16bcdc040fb